content/blog/2019-10-23-loop-unswitching/index.md

```diff
@@ -0,0 +1,252 @@
+```
+
+and
+```
+blockone:
+    x: int = const 0;
+blocktwo:
+    print x;
+```
+so we need to guarantee that these two blocks are placed next to each other when we output the new optimized program.  
+
+Finally, we connect the bypass blocks to the end of the program, which we leave intact.   
+
+The last step is to overwrite the original mapping from block name to contents with our new block dictionary and output the resulting Bril program.
+
+## Implementation Difficulties
+
+There were quite a few difficulties that arose during implementation due to the nature of the Bril programming language.
+
+As alluded to before, since Bril does not require jumps at the end of blocks, we need to be careful about reordering blocks in the final Bril program.  Furthermore, this optimization is designed to run after other optimizations, so we would ideally like to keep the majority of the program untouched to prevent overwriting.  This involves a fair amount of bookkeeping especially when these situations occur within the loop body.  
+
+Traditionally, loop unswitching operates on SSA form, and due to time restriction, I was not able to write an SSA translater.  Thus this optimization assumes a prior SSA run.  One example of why we might want SSA form is dead code elimination.  It is possible that after dead code elimination, a branch within a loop becomes independent of the loop body.
+
+Another shortcoming of this optimization is that it currently only operates on natural loops, and in particular, does not operate on nested loops.  That is, programs in the form:
+```
+b = True
+for i in range(10):
+    for j in range(10):
+        if b then:
+            do something()
+        else:
+            do something else()
+```
+will not be completely unswitched since the conditional statement lies inside of a nested for loop.  Instead, the conditional will be unswitched outside of the first for loop, which only midly decreases the numbeer of branches.  Implementing this is not too difficult to do, though it requires  additional abstractions for nested for loops.
+
+
+## Results and Evaluation 
+
+##### Results
+The results of this optimization were quite interesting.  I primarily evaluated the code on percentage difference in the number of branches.  To do this, I modified the Bril Typescript interpreter to keep track of the number of branches for each run.  For a program:
+```
+    for _ in range(maxiter):
+        prebody
+        if b:
+            ifbody
+        else:
+            elsebody
+        postbody
+```
+and its unswitched version:
+```
+    if b:
+        for _ in range(maxiter):
+            prebody
+            ifbody
+            postbody
+    else:
+        for _ in range(maxiter):
+            prebody
+            ifbody
+            postbody
+```
+the improvement comes from eliminating branches in the "if...else" decision.  Through testing, one loop unswitching produces a decrease in the number of branches of approximately 14%, on average.  
```

Comment:
On average across what? It sort of looks like this is just one program, so I would have expected there to be only one number for "reduction in branches." Or maybe you use multiple benchmarks? That would be cool to know!